src/coreclr/jit/morph.cpp

```diff
@@ -3420,7 +3615,14 @@ void Compiler::fgInitArgInfo(GenTreeCall* call)
         if (newArgEntry->isStruct)
         {
             newArgEntry->passedByRef = passStructByRef;
-            newArgEntry->argType     = (structBaseType == TYP_UNKNOWN) ? argx->TypeGet() : structBaseType;
+#if defined(TARGET_LOONGARCH64)
+            newArgEntry->argType = ((floatFieldFlags & (STRUCT_HAS_FLOAT_FIELDS_MASK ^ STRUCT_FLOAT_FIELD_ONLY_ONE)) ||
```

Comment:
I am guessing this is one more case where `getPrimitiveTypeForStruct` & `getArgTypeForStruct` return `TYP_I_IMPL`s for `struct { float, int }`s?

It would be a better workaround to fix `structBaseType` up above "fully" (it is already being fixed up for the 1-float case I see).

The actual solution would be to:
a) Store `floatFlags` in `ClassLayout` so that querying them is cheap.
b) Consult them in `getPrimitiveTypeForStruct` & `getArgTypeForStruct` so that they return correct values for `structBaseType`.
c) Come up with a `ReturnTypeDesc`-like structure but for args and let `getArgTypeForStruct` return that instead of the incomplete info it returns now (or have a "fill" mode in it where it accepts an arg entry as an in-out parameter).
d) Unify SystemV and HFAs to be on the same plan.

(Obviously, not in this change, though it would be good to have `a` and `b` in a follow-up, as that will clean up some other code too)

Reply:
> I am guessing this is one more case where `getPrimitiveTypeForStruct` & `getArgTypeForStruct` return `TYP_I_IMPL`s for `struct { float, int }`s?

Yes, you are right.

> It would be a better workaround to fix `structBaseType` up above "fully" (it is already being fixed up for the 1-float case I see).

Ok, I will do it liking this.
 
> The actual solution would be to: a) Store `floatFlags` in `ClassLayout` so that querying them is cheap. b) Consult them in `getPrimitiveTypeForStruct` & `getArgTypeForStruct` so that they return correct values for `structBaseType`. c) Come up with a `ReturnTypeDesc`-like structure but for args and let `getArgTypeForStruct` return that instead of the incomplete info it returns now (or have a "fill" mode in it where it accepts an arg entry as an in-out parameter). d) Unify SystemV and HFAs to be on the same plan.
> 
> (Obviously, not in this change, though it would be good to have `a` and `b` in a follow-up, as that will clean up some other code too)

I agree with you that there is an improvement for optimization.

Thanks